app/models/article.rb

```diff
@@ -114,18 +114,70 @@ class Article < ApplicationRecord
   after_commit :sync_related_elasticsearch_docs, on: %i[update]
   after_commit :remove_from_elasticsearch, on: [:destroy]
 
+  # The trigger `update_reading_list_document` is used to keep the `articles.search_reading_list` column updated.
+  #
+  # Its body is inserted in a PostgreSQL trigger function and it joins the following columns into a `tsvector` document:
+  # `articles.body_markdown`, `articles.cached_tag_list`, `articles.title` plus
+  # the related `organizations.name`, `users.name` and `users.username` values.
+  # Since a trigger is per table and runs before each time a SQL `INSERT` or `UPDATE` is executed in the database,
+  # we rely on Rails callbacks to propagate changes of related rows in the `organizations` and `users` tables.
+  # `Organization.update_articles` callback will eventually invoke this trigger synchronously.
+  # `Profiles::Update.call` will eventually invoke this trigger asynchronously.
+  #
+  # Please refer to https://github.com/jenseng/hair_trigger#usage in case you want to change or update the trigger.
+  #
+  # Additional information on how triggers work can be found in
+  # => https://www.postgresql.org/docs/11/trigger-definition.html
+  # => https://www.cybertec-postgresql.com/en/postgresql-how-to-write-a-trigger/
+  #
+  # Adapted from https://dba.stackexchange.com/a/289361/226575
+  trigger
+    .name(:update_reading_list_document).before(:insert, :update).for_each(:row)
+    .declare("l_org_vector tsvector; l_user_vector tsvector") do
+    <<~SQL
+      -- start with `body_markdown`, `cached_tag_list` and `title` from `articles`
+      NEW.reading_list_document :=
+        to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.body_markdown, ''))) ||
+        to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.cached_tag_list, ''))) ||
+        to_tsvector('simple'::regconfig, unaccent(coalesce(NEW.title, '')));
+
+      -- add `users.name` and `users.username`
+      SELECT to_tsvector('simple'::regconfig, unaccent(coalesce(users.name, ''))) ||
+             to_tsvector('simple'::regconfig, unaccent(coalesce(users.username, '')))
+      INTO l_user_vector
+      FROM users
+      WHERE users.id = NEW.user_id;
+
+      IF l_user_vector IS NOT NULL THEN
```

Comment:
Instead of having two `if` blocks, wouldn't the following also work (at the end of this trigger, replacing the second `if`):

```sql
NEW.reading_list_document := NEW.reading_list_document || 
  coalesce(l_user_vector, '') || 
  coalesce(l_org_vector, '');
```
Even the first  `coalesce` usage here seems a bit cautious, as by its current definition `l_user_vector` should never be `null` since you already use `coalesce` there (i.e. the wouldn't be `null` but an empty `tsvector`):

```sql
Forem_development=# \pset null 'Ø'
Null display is "Ø".
Forem_development=# select null;
 ?column?
----------
 Ø
(1 row)

Forem_development=# select to_tsvector('simple'::regconfig, '') || to_tsvector('simple'::regconfig, '');
 ?column?
----------

(1 row)
```

Reply:
That's true! I found a way to avoid JOINing the other tables altogether. I'm a bit disappointed we haven't discovered this sooner but while I was writing specs I came across `articles.cached_user_username`, `articles.cached_user_name` and `articles.cached_organization`. They are text columns and they are kept in sync by Rails callbacks. Since we already rely on callbacks for this trigger to work, I think we can simplt "vectorize" those columns and avoid the joins altogether. We'd be adding the profile image URLs to the vectorized columns but I think trading a little more text occupation for 2 fewer queries and no joins will be better :) I'm going to test this simpler trigger later

Reply:
I rewrote the trigger to use the cached values, I'm resolving this conversation :)

Reply:
Nice, I like this new version of the trigger, clever use of the cached data. If only we stored the cached data as JSON, eh? 😉